### PR TITLE
Tests of statfs/fstatfs

### PR DESCRIPTION
**Pull request by Yaxuan-w**
#### Issue Summary
- `fstatfs`: error opening file: No such file or directory.This is likely due to the issue mentioned above regarding relative paths.
- `statfs`: This runs without error, but I need to check whether the output that is returned is correct (currently the result is 0).
- Add `fstatfs` and `statfs` syscall to the test suite.

#### Change Summary
- Modify the way to get the file path in `statfs`
- Modify the way to get the file descriptor in `fstatfs`
- Add `fstatfs` and `statfs` syscall to the nondeterministic test suite

 

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

### Reproduce Instruction

- Test of `fstatfs` - `lind_project/tests/test_cases/fstatfs.c`
1. Compile `fstatfs.c` in `lindsh` with `compile tests/test_cases/fstatfs.c -o fstatfs.nexe`
2. Copy test code file to Lind file system with `lindfs cp /home/lind/lind_project/fstatfs.nexe fstatfs.nexe`
3. Copy test file to Lind file system with `lindfs cp /home/lind/lind_project/tests/test_cases/testfiles/fstatfsfile.txt /testfiles/fstatfsfile.txt`
4. Run `lind /fstatfs.nexe`

- Test of `statfs` - `lind_project/tests/test_cases/statfs.c`
1. Compile `statfs.c` in `lindsh` with `compile tests/test_cases/statfs.c -o statfs.nexe`
2. Copy test code file to Lind file system with `lindfs cp /home/lind/lind_project/statfs.nexe statfs.nexe`
3. Copy test file to Lind file system with `lindfs cp /home/lind/lind_project/tests/test_cases/testfiles/statfsfile.txt /testfiles/statfsfile.txt`
4. Run `lind /statfs.nexe`

### Relevant Information
Can also use the commented code lines to test whether syscalls can get unix system file infos.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, safeposix-rust)

### CODE REVIEW

**Comment by kuzeyardabulut** [comment]
Everything looks fine. Some suggestions:
* **tests/test_cases/fstatfs.c**: You can store the path string using **#define**, this will increase the readability of the code. For example: `#define FILE_PATH "testfiles/fstatfsfile.txt"`
* **tests/test_cases/statfs.c**: You can also use **#define** in here to store path string. 
For example: `#define FILENAME "testfiles/statfsfile.txt"`

File: tests/test_cases/fstatfs.c [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/statfs.h>
+
+int main(int argc, char **argv) {
+  // int fd = open("/dev/urandom", O_RDONLY, S_IRWXU|S_IRWXG|S_IRWXO);
```
Comment (kpan2034) [inline tests/test_cases/fstatfs.c:12]: Can remove this comment if not needed.

File: tests/test_cases/statfs.c [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/statfs.h>
+
+int main(int argc, char **argv) {
+  // char FILENAME[100] = "/dev/urandom";
```
Comment (kpan2034) [inline tests/test_cases/statfs.c:12]: Can remove this comment if not needed.

**Review by kpan2034 — CHANGES_REQUESTED** [review]
Can we also add `fstatfs` and `statfs` to the test suite before merging this?

File: tests/test_cases/fstatfs.c (hunk 1 above)
Comment (Yaxuan-w) [reply tests/test_cases/fstatfs.c:12]: I set this line to test whether fstatfs works for system file

Comment (rennergade) [reply tests/test_cases/fstatfs.c:14]: yes please remove this comment

File: tests/test_cases/fstatfs.c [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  // system file info
+  printf("Filesystem type: %lx\n", buf.f_type);
+  printf("Total blocks: %ld\n", buf.f_blocks);
+  printf("Free blocks: %ld\n", buf.f_bfree);
+  printf("Available blocks (non-root): %ld\n", buf.f_bavail);
+  printf("Total file inodes: %ld\n", buf.f_files);
+  printf("Free file inodes: %ld\n", buf.f_ffree);
+  fflush(stdout);
+  return 0;
+}
```
Comment (rennergade) [inline tests/test_cases/fstatfs.c:39]: files should end with newlines

File: tests/test_cases/statfs.c [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  // system file info
+  printf("Filesystem type: %lx\n", buf.f_type);
+  printf("Total blocks: %ld\n", buf.f_blocks);
+  printf("Free blocks: %ld\n", buf.f_bfree);
+  printf("Available blocks (non-root): %ld\n", buf.f_bavail);
+  printf("Total file inodes: %ld\n", buf.f_files);
+  printf("Free file inodes: %ld\n", buf.f_ffree);
+  fflush(stdout);
+  return 0;
+}
```
Comment (rennergade) [inline tests/test_cases/statfs.c:31]: files should end with new lines

File: tests/test_cases/statfs.c (hunk 2 above)
Comment (rennergade) [reply tests/test_cases/statfs.c:14]: Yes please

**Review by rennergade — CHANGES_REQUESTED** [review]
These tests still need the proper cosmetic changes. Also both tests are failing, so I'm not sure whats happening but that obviously needs to be fixed.

You added both of these to non-det tests, and I believe you had talked about adding py files as well. It's possible when you cleaned up the commits you removed them accidentally.

File: tests/test_cases/dettests.txt [hunk 5]
```diff
@@ -33,4 +33,4 @@ rename.c
 fchmod.c
 truncate.c
 mutex.c
-socketpair.c
+socketpair.c
```
Comment (RusherRG) [inline tests/test_cases/dettests.txt:36]: I think there is a space or newline missing. Please fix this

File: tests/test_cases/fstatfs.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+# Extract output
+# sample output: Filesystem type: 794c7630
+match2 = re. search (r'Filesystem type: (\w+)', host_result)
+if match2:
+    host_string = match2.group (1)
+# Expect host output: 794c7630
+if host_string != "794c7630":
+    print "Host info: ", host_string
+    exit(-1)
```
Comment (RusherRG) [inline tests/test_cases/fstatfs.py:28]: It is a general practice to add a newline at the end of the file
```suggestion
    exit(-1)
    
```

File: tests/test_cases/statfs.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+# Extract output
+# sample output: Filesystem type: 794c7630
+match2 = re. search (r'Filesystem type: (\w+)', host_result)
+if match2:
+    host_string = match2.group (1)
+# Expect host output: 794c7630
+if host_string != "794c7630":
+    print "Host info: ", host_string
+    exit(-1)
```
Comment (RusherRG) [inline tests/test_cases/statfs.py:28]:
```suggestion
    exit(-1)
    
```
It is a general practice to add a newline at the end of the file

File: tests/test_cases/statfs.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+if match:
+    lind_string = match.group (1)
+# Expect lind output: beefcOde
+if lind_string != "beefc0de":
+    print "Lind info: ", lind_string
+    exit (-1)
+
+# Extract output
+# sample output: Filesystem type: 794c7630
+match2 = re. search (r'Filesystem type: (\w+)', host_result)
```
Comment (RusherRG) [inline tests/test_cases/statfs.py:22]:
```suggestion
match2 = re.search(r'Filesystem type: (\w+)', host_result)
```

Could you remove spaces like this throughout the file?

File: tests/test_cases/fstatfs.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+if match:
+    lind_string = match.group (1)
+# Expect lind output: beefcOde
+if lind_string != "beefc0de":
+    print "Lind info: ", lind_string
+    exit (-1)
+
+# Extract output
+# sample output: Filesystem type: 794c7630
+match2 = re. search (r'Filesystem type: (\w+)', host_result)
```
Comment (RusherRG) [inline tests/test_cases/fstatfs.py:22]:
```suggestion
match2 = re.search(r'Filesystem type: (\w+)', host_result)
```
Could you remove spaces like this throughout the file?

**Review by RusherRG — CHANGES_REQUESTED** [review]
So per the discussion that I and @Yaxuan-w had, the value of the type of filesystem for the `native` program is `0x794c7630` which points to `OVERLAYFS_SUPER_MAGIC`, that is correct. But, the lind program returns the type as `0xbeefc0de` which is defined in [fs_calls.rs](https://github.com/Lind-Project/safeposix-rust/blob/main/src/safeposix/syscalls/fs_calls.rs#L578). 

Hence, treating tests of `statfs` and `fstatfs` as non-deterministic makes sense then.

**Comment by rennergade** [comment]
this looks good now. thanks @Yaxuan-w !

**Pull request merged**